runtime/zeitgeist/src/xcm_config/config.rs

```diff
@@ -158,6 +163,108 @@ parameter_types! {
     );
 }
 
+/// A generic warpper around implementations of the (xcm-executor) `TransactAsset` trait.
+///
+/// Aligns the fractional decimal places of every incoming token with ZTG.
+/// Reconstructs the original number of fractional decimal places of every outgoing token.
+///
+/// Important: Always use the global canonical representation of token balances in XCM.
+/// Only during the interpretation of those XCM adjustments happens.
+///
+/// Important: The implementation does not support teleports.
+#[allow(clippy::type_complexity)]
+pub struct AlignedFractionalTransactAsset<
+    AssetRegistry,
+    CurrencyIdConvert,
+    FracDecPlaces,
+    TransactAssetDelegate,
+> {
+    _phantom: PhantomData<(AssetRegistry, CurrencyIdConvert, FracDecPlaces, TransactAssetDelegate)>,
+}
+
+impl<
+    AssetRegistry: Inspect<AssetId = CurrencyId>,
+    FracDecPlaces: Get<u8>,
+    CurrencyIdConvert: Convert<MultiAsset, Option<CurrencyId>>,
+    TransactAssetDelegate: TransactAsset,
+>
+    AlignedFractionalTransactAsset<
+        AssetRegistry,
+        CurrencyIdConvert,
+        FracDecPlaces,
+        TransactAssetDelegate,
+    >
+{
+    fn adjust_fractional_places(asset: &MultiAsset) -> MultiAsset {
+        if let Some(ref asset_id) = CurrencyIdConvert::convert(asset.clone()) {
+            if let Fungible(amount) = asset.fun {
+                let mut asset_updated = asset.clone();
+                let native_decimals = u32::from(FracDecPlaces::get());
+                let metadata = AssetRegistry::metadata(asset_id);
+
+                if let Some(metadata) = metadata {
+                    let decimals = metadata.decimals;
+
+                    asset_updated.fun = if decimals > native_decimals {
+                        let power = decimals.saturating_sub(native_decimals);
+                        let adjust_factor = 10u128.saturating_pow(power);
+                        // Floors the adjusted token amount, thus no tokens are generated
+                        Fungible(amount.saturating_div(adjust_factor))
```

Comment:
Any XCM uses the global canonical representation, i.e. 18 fractional decimal places for ETH. 

Simplified:

deposit: When a reserve asset transfer with ETH is incoming, it puts the unadjusted ETH in the holding register. Then it deposits the adjusted ETH using the adjusted `deposit` function into the target pallet and account. Then it subtracts the unadjusted value from the holding register.

withdraw: When a reserve asset transfer with ETH is sent, it withdraws the adjusted ETH using the adjusted `withdraw` function from the target pallet and acount and afterwards puts the unadjusted ETH in the holding register (since the XCM contains the global canonical representation, in this case 18 fractional decimal places).

deposit: XCM in (unadjusted ETH) -> holding register (unadjusted) -> tokens (deposit, adjusted) -> update holding register (unadjusted)
withdraw: XCM out (unadjusted ETH) -> tokens (withdraw, adjusted) -> holding register (unadjusted)

Reply:
The trick here is that we demand the global canonical representation for any token. Simplified, if 1 ETH enters, the XCM contains the amount `10^18`. If an XCM is created on Zeitgeist to send ETH, the XCM contains the amount `10^18` as value, although internally it is stored as `10^10`.

Reply:
I see. So the point that I missed is that any XCM message either `withdraw` or `deposit` is (hopefully) always constructed with the external decimal representation (not our internal ten base representation). 

I also missed the point in general that your `AlignedFractionalTransactAsset` is only meant for the internal representation to get data from the XCM into the tokens pallet. `AlignedFractionalTransactAsset` is not meant to get the asset amount out to the external representation, since the external representation is always the value inside the XCM.

It reminds me about a one-way hash function into the internal representation only.